feat(chat): 채팅 메시지 송수신 기능 구현

j1suk1m opened the pull request:
## 📌 목적
- 채팅 메시지를 실시간으로 주고받는 기능을 구현합니다.

## ✅ 변경 사항
- HTTP REST API와의 명확한 구분을 위해 WebSocket handshake URL을 `/chat` -> `/ws/chat`으로 변경했습니다.

## 🧪 테스트 방법
- [초기 설정](https://github.com/prgrms-be-adv-devcourse/beadv1_1_hexagon_BE/issues/12)의 테스트 방법과 동일합니다.
- 서비스 단위 테스트인 ChatMessageServiceTest를 실행합니다.

## 📎 참고 사항
- 현재 요청 바디에 메시지 송신자의 `code`를 직접 포함하고 있는데, 추후 인터셉터를 구현해 보안을 강화할 예정입니다.
- 엔티티의 `LocalDateTime` 필드를 `Instant`로 변경했습니다.

## 📝 제출자 검토 리스트

- [x] 실행 시 오류가 없나요?
- [x] 테스트 전부 통과 했나요?
- [x] 코드 컨벤션을 모두 지켰나요?
- [x] 브랜치가 develop에 merge 요청을 보냈나요?

## 📝 검토자 체크 리스트

> [!IMPORTANT]
> 직접 모든 부분에 대해 검토하고 멘션이나 이모티콘 남겨주세요

- [ ] 요청을 받은 후 하루 이내에 피드백을 보내주세요
- [ ] 본인이 검토해야하는 PR 피드백을 주세요.
- [ ] 브랜치가 develop에 merge 요청을 보냈나요?

j1suk1m committed:
chore(config): WebSocket handshake URL 변경

- REST API와 명확히 구분하기 위해 `/chat`를 `/ws/chat`으로 변경
- HTML에 변경 사항 반영

j1suk1m committed:
feat(dto): 채팅 메시지 전송 요청 및 응답 DTO 구현

- id와 code를 명확히 구분하기 위해 요청 DTO의 senderId를 senderCode로 변경
- HTML에 변경 사항 반영

j1suk1m committed:
feat(chat): 채팅 메시지 송수신 기능 구현

- `@MessageExceptionHandler`를 이용한 WebSocket 전용 예외 처리
- 채팅 메시지 저장 시 해당 메시지가 속하는 채팅방의 updatedAt 갱신 처리

haeun9988 commented on communication-service/src/main/java/com/example/communicationservice/controller/ChatMessageController.java:
```diff
@@ -1,21 +1,46 @@
+     * 3. 해당 토픽 구독자에게 메시지 전송
+     */
     @MessageMapping("chat.send")
     public void handleChatMessage(ChatMessageSendRequest request) {
```
updatedAt 갱신을 별도의 비동기/배치 작업으로 분리하거나 ChatMessage 저장 시점을 활용하도록 ChatRoom의 최신 메시지 생성 시점을 쿼리하여 사용하는 편이 조금 더 병목 현상을 덜 유발 시킬 것 같은데 어떠신가요?!

j1suk1m replied:
좋은 말씀 감사합니다! 여러 대안을 고민해보고 전후 성능 비교해서 문서에 작성해보겠습니다👀

ZZAMBAs commented on communication-service/src/main/java/com/example/communicationservice/controller/ChatMessageController.java:
```diff
@@ -1,21 +1,46 @@
+
+    // WebSocket 전용 예외 처리
+    @MessageExceptionHandler(ChatRoomException.class)
+    @SendToUser("/queue/errors")
```
채팅 사용자들은 해당 URL을 구독하고 있다가 에러 메시지를 받는 걸까요?

j1suk1m replied:
맞습니다. 프론트엔드에서 명시적으로 구독해줘야 합니다!

haeun9988 commented on communication-service/src/main/java/com/example/communicationservice/controller/ChatMessageController.java:
```diff
@@ -1,21 +1,46 @@
+    }
+
+    // WebSocket 전용 예외 처리
+    @MessageExceptionHandler(ChatRoomException.class)
```
@MessageExceptionHandler이 WebSocket 전용 예외 처리군요 좋은 방법입니다!

haeun9988 reviewed (approved):
젙체적으로 보기 깔끔한 코드입니다! LGTM!

ZZAMBAs reviewed (approved):
궁금한 점만 아래 하나 남겼습니다.

programofktw commented on communication-service/src/main/java/com/example/communicationservice/service/ChatMessageService.java:
```diff
@@ -53,4 +56,37 @@ public ChatMessageListReadResponse findMessagesByRoomId(String roomId, String cu
+        ChatRoom chatRoom = chatRoomRepository.findById(roomId)
+            .orElseThrow(() -> new ChatRoomException(ResponseDtoStatus.CHATROOM_NOT_FOUND));
+
+        // 메시지 송신자가 해당 채팅방의 참여자인지 확인
```
인증은 되어있지만 권한이 없는 것이니 UNAUTHORIZAED 대신 FORBBIDEN은 어떨까요!

j1suk1m replied:
좋은 조언 감사합니다! 순간 인증 인가를 반대로 생각했네요🤯

programofktw commented on communication-service/src/main/java/com/example/communicationservice/entity/ChatRoom.java:
```diff
@@ -28,6 +25,7 @@ public class ChatRoom {
     private LocalDateTime createdAt;
 
     @LastModifiedDate
+    @Setter
```
어째서 Setter를 쓰셨나 찾아보았는데
MongoDb의 경우 LastModifiedDate 가 작동하기 위해서는 Setter가 필요하군요!

j1suk1m replied:
꼭 setter가 있어야만 동작하는 것은 아닌 것으로 알고 있습니다!
저는 메시지가 채팅방에 저장될 때마다 채팅방의 updatedAt을 갱신하기 위해 사용했습니다👀

programofktw reviewed (approved):
프론트 코드까지 고생이 많으셨습니다!

j1suk1m committed:
fix(service/dto): LocalDateTime을 Instant로 변경

- 테이블 설계 시 논의했던 사항 반영

j1suk1m committed:
fix(status): 권한이 없는 상황을 나타내는 예외 코드 수정

- 기존의 CHATROOM_UNAUTHORIZED는 인증 실패에 가까움
- 인가 오류를 명확히 표현하기 위해 CHATROOM_FORBIDDEN으로 변경

j1suk1m committed:
test(chat): 채팅방 메시지 저장 로직 서비스 단위 테스트 추가

j1suk1m committed:
chore(import): 사용하지 않는 import 제거
